Allow user to assign value to rubric choices #4746

unyoungwax opened the pull request:
Fixes #4746

Edit: added screenshots
InstructorFeedbackEditPage:
![question](https://cloud.githubusercontent.com/assets/10228385/14143547/6c9fa340-f6be-11e5-8b82-c27169bb6457.jpg)

InstructorFeedbackResultsPage:
![result](https://cloud.githubusercontent.com/assets/10228385/14143555/76d9dcae-f6be-11e5-959b-20406bd9157c.jpg)

kanghj commented:
@unyoungwax, post screenshots of your changes.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -52,17 +55,26 @@ public boolean extractQuestionDetails(
+                try {
+                    rubricWeights.add(Integer.parseInt(weight));
+                } catch (NumberFormatException e) {
+                    rubricWeights.add(0);
```
I don't think catching this is a good idea. I think we should throw an error to the user and show an error message to the user instead if they are supplying non-integer values

unyoungwax replied:
Since the method `extractQuestionDetails` does not throw any exception, I will check it in `validateQuestionDetails`.

kanghj replied:
Yup, then replace this <code>add(0)</code> to an Assumption.fail

unyoungwax replied:
If `Assumption.fail` is used in the catch-block, the error page will be shown when the weight cannot be parsed. I tried using `add(null)` and then check for null in `validateQuestionDetails`, it works well.

kanghj replied:
Right, I forgot <code>validateQuestionDetails</code> is called only after the question details have been constructed. I'm not very sure what to feel about storing <code>null</code> as a value in the list.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -737,5 +800,18 @@ public String getQuestionTypeChoiceOption() {
 
+    private boolean hasAssignedWeights() {
 
+        if (rubricWeights == null || rubricWeights.size() != numOfRubricChoices) {
```
We can remove the test for null, since its initialised to an empty list

kanghj commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -1550,6 +1571,36 @@ function highlightRubricCol(index, questionNumber, highlight) {
+    var areWeightsAssigned = false;
+
+    $('#questionTable' + questionNumber).find('input[id^="rubricWeight"]').each(function() {
+        if ($(this).val() != 0) {
```
<code>!==</code>

unyoungwax replied:
`val()` returns a string, should I change the line to `$(this).val() !== '0'`?

kanghj replied:
Perhaps try to parseInt ?

kanghj commented:
On the screenshot for the edit page, did you miss the comment on the issue page to place the "assign weights" button on the top-left cell? Or is there a reason that the current position would work better?

unyoungwax commented:
I didn't notice it previously, I will move the "assign weights" button to the top-left cell in the next few commits.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -52,17 +55,26 @@ public boolean extractQuestionDetails(
             String choice = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + i);
-            if(choice != null) {
+            String weight = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT + "-" + i);
+            if(choice != null && weight != null) {
```
Just to check, when will this condition ever be false?

unyoungwax replied:
The condition will be false if the weight or choice is not found in the `requestParameters` (probably due to the user editing the html).

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -364,6 +380,11 @@ public String getNewQuestionSpecificEditFormHtml() {
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
```
@damithc, just to check, if the instructor does not assign weights, do we want to give the rubrics choices default weights, or should not display the weights (done in this pr)?

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -460,19 +481,27 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
             String tableHeaderCell = 
                     FeedbackQuestionFormTemplates.populateTemplate(tableHeaderFragmentTemplate,
-                            "${rubricChoiceValue}", rubricChoiceValue);
+                            "${rubricChoiceValue}", rubricChoiceValue + (hasAssignedWeights() ? "(" + rubricWeight + ")" : ""));
```
Split this line into 2

unyoungwax commented:
I have moved the "assign weights" button to the top-left cell.
![question](https://cloud.githubusercontent.com/assets/10228385/14204276/b03a4a9c-f835-11e5-9498-7ef8ee6942c4.jpg)

I have also made some changes so that a warning message is shown when the weights are not integer.
![status](https://cloud.githubusercontent.com/assets/10228385/14204280/ba1befde-f835-11e5-90b6-1d9eca3cf10f.jpg)

ashrayjain commented:
@unyoungwax Please revert the permissions change to the file ```./gradlew```. This is causing the Travis build for this PR to break. Thanks! :)

damithc commented:
@unyoungwax No reason to restrict weights to integers right? It should be fine as long as they are numbers?

unyoungwax commented:
Yes, we can assign floating-point numbers to the weights. How many decimal places should be allowed for the weights?

damithc commented:
I guess 2 decimal points is reasonable

unyoungwax commented:
I have made some changes to allow assigning floating-point numbers to weights.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -520,10 +557,8 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
         
         // Initialize percentage frequencies and average value
-        float[][] percentageFrequencyOrAverage = new float[fqd.numOfRubricSubQuestions][];
+        double[][] percentageFrequencyOrAverage = new double[fqd.numOfRubricSubQuestions][fqd.numOfRubricChoices + 1];
```
Why <code>+ 1</code>? You can add a comment in the code to explain why

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -605,6 +662,11 @@ public String getQuestionResultStatisticsCsv(
             }
+
+            if (hasAssignedWeights) {
+                csv.append("," + dfAverage.format(rubricStats[i][numOfRubricChoices]));
```
This <code>rubricStats</code> variable can contain more than one type of value?

unyoungwax replied:
`rubricStats` stores the result of `calculateRubricStats`, an array containing the percentage frequencies and averages.

kanghj replied:
Ok, got it. We can leave it for now, but it looks like it should be done as 2 methods, one for the percentage frequencies and one for average.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -737,5 +804,18 @@ public String getQuestionTypeChoiceOption() {
+            return false;
+        }
+
+        for (Double weight : rubricWeights) {
```
Let's just use <code>double</code> here.

kanghj commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -1550,6 +1572,36 @@ function highlightRubricCol(index, questionNumber, highlight) {
+ * of the input boxes for rubric weights
+ */
+function bindAssignWeightsButtons() {
+    $('body').on('click', 'a[id^="rubricAssignWeightsLink"]', function(e) {
```
Why is this needed? Doesn't using the selector normally work?

unyoungwax replied:
It is easier to maintain the code using this method instead of attaching event handlers directly to the buttons. See #3998

kanghj commented:
On the edit page, when the "edit" on the question hasn't been clicked, the "assign weights" button is missing. It should instead be there, but disabled, to be consistent with the rest of our UI.

kanghj commented:
I think we have to add a tooltip to the "assign weights" button and the weights input to describe what the weights are. Starting with default values of 0s for all the options may not be good too.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -14,6 +14,7 @@
 public class FeedbackRubricQuestionDetails extends FeedbackQuestionDetails {
     public int numOfRubricChoices;
     public List<String> rubricChoices;
+    public List<Double> rubricWeights;
```
Have you tested with legacy data? Will there be problems with old rubrics questions without <code>rubricsWeights</code>?

unyoungwax replied:
I am not sure how to test it correctly. I have no problems editing and showing the results of the rubric question in "Session with different question types"(one of the default sessions in the dev server).

kanghj commented:
Btw, even though it's not merged into master yet, do take note of #5024

unyoungwax commented:
If the user clicks the "assign weights" button to hide the weights, should the weights be reset to default or restored to the previous version? Either way we need to change the text of the button to "reset weights" or "restore weights" and add tooltips to explain it.

unyoungwax commented:
Since the difference between the number of rubric choices and the number of weights is used to determine if there is an empty weight, both values must be equal to edit or add a question successfully. If the user does not assign any weights and the default weights are not all zeros, it is hard to tell whether the user wants to assign weights or not.

unyoungwax commented:
I made the "assign weights" button to be similar to the "add row" and "add column" button. The buttons are not shown when the "edit" button has not been clicked. Does the "assign weights" button need to have a different behavior?

damithc commented:
Perhaps a check box 'weighted values' is more appropriate than a button? It should be unticked by default.  
Tick the check box --> shows the text boxes. We can auto-fill some default values (not urgent, can be done later)
Untick the check box --> hides the text boxes. Values are preserved but not used. (if resetting is easier, we can reset for now. Preserving can be done in a future version).

unyoungwax commented:
![question](https://cloud.githubusercontent.com/assets/10228385/14231527/58f54662-f9b9-11e5-9f18-d51ec63fd9aa.jpg)
Is this okay?

damithc commented:
Placement is fine. I'm not sure 'values' is the right term though. 'Weighted columns'?

unyoungwax commented:
I will rename "Weighted values" to "Weighted columns". I am not sure if the users know that the term "choices" means rubric choices. I will change the tooltip to "Assign weights to columns to calculate the average".

damithc commented:
`... calculate average` --> `... calculate statistics`?

unyoungwax commented:
I have change the input type of "assign weights" from button to checkbox. When the checkbox is ticked, the text boxes are shown and auto-filled with zeros. When the checkbox is unticked, the text boxes are hidden. The weights are not preserved.

kanghj commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -290,7 +291,7 @@
         public static final String RUBRIC_ERROR_EMPTY_SUB_QUESTION = "Sub-questions for "+Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_INVALID_CHOICE = "An invalid choice was chosen for the " + Const.FeedbackQuestionTypeNames.RUBRIC + ".";
-        
+        public static final String RUBRIC_ERROR_INVALID_WEIGHT = "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.RUBRIC + " must be valid numbers with precision up to 2 decimal places.";
```
I think we can just give a generic failure message here for this.

kanghj commented:
Perhaps the tooltip should be rephrased to "... to calculate statistics." or "... for calculating statistics.", instead of "... to calculate the statistics."? 

Code looks good to me, but there are some merge conflicts to be resolved now.

unyoungwax commented:
I have resolved the merge conflict and also change the tooltip of  "assign weights" as suggested.

damithc commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -44,20 +50,23 @@ public boolean extractQuestionDetails(
             FeedbackQuestionType questionType) {
         String numOfRubricChoicesString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_NUM_COLS);
         String numOfRubricSubQuestionsString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_NUM_ROWS);
+        String hasAssignedWeightsString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED);
```
This line, and a few other lines are too long. wrap?

unyoungwax replied:
I will wrap the lines like this
```java
String hasAssignedWeightsString = HttpRequestHelper.getValueFromParamMap(
        requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED);
```

damithc commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -66,7 +75,33 @@ public boolean extractQuestionDetails(
+            try {
+                rubricWeights.add(Double.parseDouble(weight));
+            } catch (NumberFormatException e) {
+                // Do not add weight to rubricWeights if the weight cannot be parsed
```
Is there some mechanism to prevent/warn users from adding unparsable weights?

unyoungwax replied:
The `type` of the input boxes for the weights is `number`. If the user adds unparsable weights, the browser will prevent form submission and show a warning message. However, it is still possible for the user to assign empty weights, therefore we need to check them in `FeedbackRubricQuestionDetails`.

![error](https://cloud.githubusercontent.com/assets/10228385/14308885/ad612b7c-fc0b-11e5-8b42-70710d684e44.jpg)
The browser that I am using is Chrome.

damithc commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -1561,6 +1581,25 @@ function highlightRubricCol(index, questionNumber, highlight) {
+ *            the question number of the feedback question
+ * @returns {Boolean} true if the weights are assigned by the user, otherwise false
+ */
+function hasAssignedWeights(questionNumber) {
```
@kanghj I think we still have cases of duplicate question numbers. Can it break things here? Should we use a more trustworthy key to distinguish questions?

kanghj replied:
Yes, it will break things here. Most of the existing logic and behaviour in this file relies on question numbers though, we should make a new issue to change the existing logic in this file and in the InstructorFeedbackEditPageData to use the questionId instead.

damithc replied:
I don't think we can release this feature if it has the potential to break in production. Hence cannot merge. :-(

damithc commented:
Don't we need to update instructor help page as well?

damithc commented:
What happens if a an instructor adds weights to (or modifies weights of) a rubric question after the session has collected some responses?

unyoungwax commented:
If an instructor assigns or modifies weights, existing responses will not be affected (they will not be deleted).

unyoungwax commented:
I plan to update the instructor help page as follows:

![question](https://cloud.githubusercontent.com/assets/10228385/14310734/24bdb0dc-fc16-11e5-99fc-0b8471f46c8e.jpg)

![statistics](https://cloud.githubusercontent.com/assets/10228385/14310736/294223a4-fc16-11e5-944b-e0305c6a0f69.jpg)

Is it okay?

damithc commented:
The Weight text box can have a `Weight:` label in front? It doesn't need to be as wide as the text box above it.
How are weights shown in the submission page, and the student results page?

unyoungwax commented:
The weights are not shown in the submission page. The statistics table is the same for both student results page and instructor feedback results page, i.e. the weights are shown only if the weights are assigned.

damithc commented:
Can post screenshots of submission and result pages? If the student didn't see weights during submission, won't she be surprised to see them in results?

kanghj commented:
@damithc, I think we can do the submission page in a different issue?

unyoungwax commented:
Submission page:
![student_submit](https://cloud.githubusercontent.com/assets/10228385/14314563/8698b64e-fc2a-11e5-8d63-170df274df3b.jpg)

Student results page:
![student_result](https://cloud.githubusercontent.com/assets/10228385/14314578/99ae949c-fc2a-11e5-9e0d-5e747e7f61de.jpg)

Instructor results page:
![instructor_result](https://cloud.githubusercontent.com/assets/10228385/14314588/a5f6bea0-fc2a-11e5-9636-2a1a1c4a1829.jpg)

I think a student will be confused if he/she sees the weights because he/she does not know what the numbers mean. If the student see the weights in the results page, he/she can know what the numbers mean because there is an "average" column in the table.

unyoungwax commented:
For the weight labels, is the following okay?
![label](https://cloud.githubusercontent.com/assets/10228385/14315496/36e4535a-fc30-11e5-89db-975ca673acd4.jpg)

damithc commented:
Um... doesn't look nice. :-p Any other options?

unyoungwax commented:
I think this is the better way to do it because the content of the table cell will not overflow when there are too many columns. Perhaps we can set the placeholder for the text boxes to "weight" instead of auto-filling the text boxes with zeros?

damithc commented:
@kanghj I think it is better to do both student and instructor sides together. But enhancements can be done later.
@unyoungwax try to show the word `Weight:` where it is possible. e.g. in student result form

unyoungwax commented:
I tried to place the text boxes in a new row.

Before and after clicking on the "weights" checkbox:
![hidden](https://cloud.githubusercontent.com/assets/10228385/14341299/2de986e2-fcc3-11e5-92c3-3d69cb65584e.jpg)
![row](https://cloud.githubusercontent.com/assets/10228385/14341294/27055144-fcc3-11e5-9f14-ca67d4055b09.jpg)

unyoungwax commented:
The weight labels in student and instructor results page, are they okay?
![statistics](https://cloud.githubusercontent.com/assets/10228385/14342539/ebb9b904-fccc-11e5-9a02-58c8b2fde547.jpg)

damithc commented:
Before clicking, show only one row and just the `[ ] Weights`.
After clicking, show both rows and both `Choices --> ` and `[ ] Weights -->`
Yes, the instructor page looks fine. `(...)` part can be made non-bold, but it can be left as a `contributor` issue for future devs.

unyoungwax commented:
I have added a "weights" row for rubric question in InstructorFeedbackEditPage. "Weight" labels are also added in the student and instructor results page as well as in question statistics csv.

Before and after clicking on the "weights" checkbox:
![hidden](https://cloud.githubusercontent.com/assets/10228385/14355159/9444ba52-fd12-11e5-9c3c-4edcb959e123.jpg)
![row](https://cloud.githubusercontent.com/assets/10228385/14355170/9ac6a278-fd12-11e5-9bfe-4dac6e550eb7.jpg)

The results page:
![statistics](https://cloud.githubusercontent.com/assets/10228385/14355191/af058bfa-fd12-11e5-9298-b2e7aa975bd1.jpg)

damithc commented:
Minor things: there should be a space between the check box and the label. Adding arrows (e.g. Choices --> , Weights --> ) can clarify what the label is referring to.

unyoungwax commented:
Arrows have been added to the "Choices" and "Weights" table cells.

![hidden](https://cloud.githubusercontent.com/assets/10228385/14380138/905b4ec6-fdb0-11e5-84d2-5daef354b258.jpg)
![row](https://cloud.githubusercontent.com/assets/10228385/14380141/94b084a0-fdb0-11e5-870e-b7e0d811b2cf.jpg)

damithc commented:
@kanghj can have another look?

kanghj commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -1446,8 +1457,26 @@ function addRubricCol(questionNumber) {
+
+    // Insert after last <th>
+    var lastWeightCell = $('#rubricWeights-' + questionNumber + ' th:last');
+    $(rubricWeightFragment).insertAfter(lastWeightCell);
```
Just to check, although I am not sure what the value of calling <code>disallowNonNumericEntries</code> is, I think to be consistent, we need to call it here too

damithc replied:
What's the decision on this one?

damithc commented:
@kanghj I feel this cannot be merged because it can break during production due to question number issue.  At least there should be an easy way for me to rectify the issue if it happens. What do you think? I occasionally get 'wrong question number' warnings, so that issue is not completely resolved yet and can affect this feature.

kanghj commented:
We can put this onHold then, until I complete #5208. I don't really think a fix local to this pr is the best way to go about fixing this problem

damithc commented:
Sure, let's do that. If it's too big to do now, we can do it after exams.

damithc commented:
@unyoungwax can finish this?

unyoungwax commented:
I will work on this as soon as #5208 is resolved.

kanghj commented:
@unyoungwax, we have resolved #5208.

unyoungwax commented:
I have resolved the conflicts with the master branch.

kanghj commented:
@unyoungwax, The tests are failing though, do you have time to fix it?

unyoungwax commented:
I ran all the tests on my computer and they passed. The tests that fail in Travis are not related to this pr (`TimeoutException` in `InstructorCourseStudentDetailsEditPageUiTest` and `NoSuchElementException` in `StudentProfilePageTest`).

kanghj commented:
Ok, I'll restart the build

unyoungwax commented:
I have resolved the conflicts with the master branch.

wkurniawan07 commented:
@damithc this one is ready for your review.

damithc commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -316,7 +317,7 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
         public static final String RUBRIC_ERROR_EMPTY_SUB_QUESTION = "Sub-questions for " + Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_INVALID_CHOICE = "An invalid choice was chosen for the " + Const.FeedbackQuestionTypeNames.RUBRIC + ".";
-        
+        public static final String RUBRIC_ERROR_INVALID_WEIGHT = "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.RUBRIC + " must be valid numbers with precision up to 2 decimal places.";
```
Line too long

damithc commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -1542,6 +1575,59 @@ function highlightRubricCol(index, questionNumber, highlight) {
+}
+
+/**
+ * Move the "weights" checkbox to the weight row if it is checked, otherwise
```
Moves

damithc commented:
Minor cosmetic changes requested. In particular, use line wrapping where the github diff view requires horizontal scrolling.

unyoungwax commented:
I have performed the requested cosmetic changes and also made some changes according to the static analysis.

wkurniawan07 commented:
@kanghj let's review (and hopefully merge) this one quickly before we introduce more changes that might conflict with this PR again.

wkurniawan07 commented:
@damithc ready for your review.

wkurniawan07 commented:
@kanghj I'm merging this now, before more stuffs are added and rendered this PR conflicted again.

@unyoungwax thanks a lot for your perseverance! :+1: :+1:
